opentelemetry-sdk/src/metrics/internal/sum.rs

```diff
@@ -43,26 +95,41 @@ impl<T: Number<T>> ValueMap<T> {
             self.no_attribute_value.add(measurement);
             self.has_no_value_attribute_value
                 .store(true, Ordering::Release);
-        } else if let Ok(mut values) = self.values.lock() {
-            let size = values.len();
-            match values.entry(attrs) {
-                Entry::Occupied(mut occupied_entry) => {
-                    let sum = occupied_entry.get_mut();
-                    *sum += measurement;
-                }
-                Entry::Vacant(vacant_entry) => {
-                    if is_under_cardinality_limit(size) {
-                        vacant_entry.insert(measurement);
-                    } else {
-                        values
-                            .entry(STREAM_OVERFLOW_ATTRIBUTE_SET.clone())
-                            .and_modify(|val| *val += measurement)
-                            .or_insert(measurement);
-                        global::handle_error(MetricsError::Other("Warning: Maximum data points for metric stream exceeded. Entry added to overflow.".into()));
-                    }
-                }
+            return;
+        }
+
+        // Determine the bucket index for the attributes
+        let bucket_index = Self::hash_to_bucket(&attrs);
+        let mut bucket_guard = self.buckets[bucket_index].lock().unwrap();
+        if let Some(bucket) = &mut *bucket_guard {
+            // if the attribute is present in bucket, lets update the value and return
+            // (we need to be fast here, as this is most common use-case)
+            if let Some(value) = bucket.get_mut(&attrs) {
+                *value += measurement;
+                return;
             }
         }
+
+        // if the attribute is not present in bucket, lets unlock the bucket
+        drop(bucket_guard);
+        // Attempt to first increment the total unique entries if under limit.
+        let under_limit = self.try_increment();
+        // Determine the bucket index for the attributes
+        let (bucket_index, attrs) = if under_limit {
+            (bucket_index, attrs) // the index remains same
+        } else {
+            // TBD - Should we log, as this can flood the logs ?
+            (OVERFLOW_BUCKET_INDEX, STREAM_OVERFLOW_ATTRIBUTE_SET.clone())
+        };
+
+        // Lock and update the relevant bucket
+        let mut final_bucket_guard = self.buckets[bucket_index].lock().unwrap();
```

Comment:
this wont work, because by the time we reach and acquire this lock, collect() may have triggered and wiped the hashmap clean, and another updates could have occurred as well making our assumption that this attribute is not-an-overflow invalid.

let under_limit = self.try_increment();
It is entirely possible that this thread loses its CPU right after the above statement is executed. When it gets a chance to execute again, under_limit could have changed already. We need to do this whole thing atomically.

Reply:
There are two changes in the design
 - We increment before the actual insert.
 - The increment i.e, try_increment() is atomic.

There is a narrow window where a measurement with new set of attributes go to overflow index even when there is empty index (this can happen during collect), which should be fine. However, there won't be a scenario where the number of data-points exceed the cardinality limit.

Reply:
And thanks for the thorough review of this PR, thinking of all possible scenarios and edge conditions :)

Reply:
take this example:
thread1 is executing update with attributes (foo=bar)., the current total_unique is 100, with limit as 2000
say bucket_index is 4
take lock
(foo=bar) is not present. 
drop lock
try_increment is called which increments total_unique to 101,  which is within the limit., so this thread will go and update bucket_index 4.
====thread1 lost its CPU===

// in other threads
collect() occurred and it reset total_unique to 0
other updates() occurred and take total_unique to 2000 (limit hit). None of these updates() were having (foo=bar) as attribute. 
At this stage (foo=bar) should be going into overflow.

====thread1 got back its CPU===
It takes lock to update bucket_index 4 with (foo=bar). But this is not correct. (foo=bar) should be going to overflow only.

Reply:
> collect() occurred and it reset total_unique to 0

It won't cleanup to 0, but will only adjust(decrement) according to the number of entries read.

Reply:
collect() read and drained the entire hashmaps. Won't it hit 0 then?

Reply:
In the above scenario, when the number of entries in hashmap is 100, while the unique_count is 101 (for foo=bar), if collect happens before foo=bar is inserted, the count will get reset to 1 (not 0), and then the insert happens after/during the collect.

Reply:
take the example where 5 threads try to update with (foo=bar) which is not existing currently. The current unique_count is 1999, and limit is 2000.
All 5 threads realize entry not found, so they all attempt to increment the unique_count. 
first thread succeeds, and attempts to store into correct bucket but remaining 4 threads will attempt to put foo=bar into overflow. I believe this is the part which you said might be okay?

However, consider the same scenario, but the current unique_count was just 10.
All 5 threads realize entry not found, so they all attempt to increment the unique_count.
All threads would succeed, so unique_count reaches 15. (though it should actually be 11)
One threads would then **insert** the hashmap and **rest** would update the hashmap.

Collect() runs, and it drains the map, but when it drains (foo=bar), it'll reduce unique_count by 1 (from 15 to 14). So we now have 4 wasted entries. No matter how many collect() runs afterwards, we'll never reclaim that wasted spot... If we run stress test long enough (may weeks), we might see all entries going to overflow, even though there is plenty of space.

Please confirm if the above is correct.